rust/openvasctl/src/ctl.rs

```diff
@@ -1,83 +1,198 @@
 use std::{
-    collections::HashMap,
+    collections::{HashMap, VecDeque},
     process::Child,
     sync::{Arc, Mutex},
+    time::Duration,
 };
 
+use sysinfo::System;
+use tokio::{sync::mpsc, task};
+
 use crate::{
     cmd::{self, check_sudo},
+    config::Config,
     error::OpenvasError,
 };
 
 #[derive(Default)]
 pub struct ScanContainer {
+    queue: VecDeque<models::Scan>,
     init: HashMap<String, ()>,
     running: HashMap<String, Child>,
 }
 
-pub struct OpenvasController {
-    scans: Arc<Mutex<ScanContainer>>,
+impl ScanContainer {
+    /// Add a scan into the init phase
+    fn dequeue(&mut self) -> models::Scan {
+        let next = self.queue.pop_front().unwrap();
+        self.init.insert(next.scan_id.clone().unwrap(), ());
+        next
+    }
+
+    /// Removes a scan from init and add it to the list of running scans
+    fn add_running(&mut self, id: String, sudo: bool) -> Result<bool, OpenvasError> {
+        if self.init.remove(&id).is_none() {
+            return Ok(false);
+        }
+        let openvas = cmd::start(&id, sudo, None).map_err(OpenvasError::CmdError)?;
+        self.running.insert(id, openvas);
+        Ok(true)
+    }
+
+    fn remove_queue(&mut self, id: &String) -> bool {
+        if let Some(index) = self
+            .queue
+            .iter()
+            .position(|x| x.scan_id == Some(id.to_owned()))
+        {
+            self.queue.remove(index);
+            return true;
+        }
+        false
+    }
+
+    /// Remove a scan from the init phase
+    fn remove_init(&mut self, id: &String) -> bool {
+        self.init.remove(id).is_some()
+    }
+
+    /// Remove a scan from the list of running scans and returns the Child process
+    fn remove_running(&mut self, id: &String) -> Option<Child> {
+        self.running.remove(id)
+    }
+
+    fn num_running(&self) -> usize {
+        self.init.len() + self.running.len()
+    }
+}
+
+type Scans = Arc<Mutex<ScanContainer>>;
+
+pub enum ScanCommand {
+    StartScan(models::Scan),
+    StopScan(String),
+}
+
+pub struct Scheduler {
+    scans: Scans,
     sudo: bool,
+    config: Config,
 }
 
-impl OpenvasController {
+impl Scheduler {
```

Comment:
maybe it makese sense to move the scheduler to plugin scheduler and move everything openvas specific to openvasctl struct? That way we could utilize the scheduler with the rust impl.